taichi/backends/cuda/cuda_context.cpp

```diff
@@ -60,39 +66,38 @@ void CUDAContext::launch(void *func,
     profiler->start(task_name);
   if (gridDim > 0) {
     std::lock_guard<std::mutex> _(lock);
-    check_cuda_error(cuLaunchKernel((CUfunction)func, gridDim, 1, 1, blockDim,
-                                    1, 1, 0, nullptr, arg_pointers.data(),
-                                    nullptr));
+    driver.launch_kernel(func, gridDim, 1, 1, blockDim, 1, 1, 0, nullptr,
+                         arg_pointers.data(), nullptr);
   }
   if (profiler)
     profiler->stop();
 
   if (get_current_program().config.debug) {
-    check_cuda_error(cudaDeviceSynchronize());
-    auto err = cudaGetLastError();
-    if (err) {
-      TI_ERROR("CUDA Kernel Launch Error: {}", cudaGetErrorString(err));
-    }
+    driver.stream_synchronize(0);
   }
 }
 
 CUDAContext::~CUDAContext() {
+  // TODO: restore this?
   /*
-  check_cuda_error(cuMemFree(context_buffer));
+  CUDADriver::get_instance().cuMemFree(context_buffer);
   for (auto cudaModule: cudaModules)
-    check_cuda_error(cuModuleUnload(cudaModule));
-  check_cuda_error(cuCtxDestroy(context));
+      CUDADriver::get_instance().cuModuleUnload(cudaModule);
+  CUDADriver::get_instance().cuCtxDestroy(context);
   */
 }
 
 CUDAContext &CUDAContext::get_instance() {
-  if (!instance) {
-    instance = std::make_unique<CUDAContext>();
+  auto tid = std::this_thread::get_id();
+  if (instances.find(tid) == instances.end()) {
```

Comment:
Please see my comments on `CUDADriver` on the singletons. Could you move `instances` as a local `static` variable here? Also i think you need a `static std::mutex` to guard the usage..